### ISSUE TITLE
Warn about master glyphset/Unicode incompatiblities

### ISSUE THREAD

**Issue by simoncozens** [googlefonts/fontmake#733]
Here's a fun fact I've just discovered: in a designspace -> variable font workflow, what happens if the master UFOs have different glyphsets or different Unicode codepoint assignments?

The answer seems to be that the first master file mentioned in the designspace file wins: it decides the glyphset and the Unicode codepoints. Then all the glyphs are collected across the masters *by name*, even if their codepoint assignments are different. Glyphs not in the first master's glyphset are ignored.

I'm not suggesting there's a better way to do this because with UFO -> variable font you are essentially trying to merge a bunch of independent font files. But fontmake provides *no warnings at all* that anything is unusual. We check for contour incompatibility, but nothing else.

(I don't know if this should be considered a fontmake problem or a ufo2ft problem or a fontTools.varLib.merge problem...)

**Comment by anthrotype** [comment]
> the first master file mentioned in the designspace file wins

I think it's not the first, but the _default_ source (implicitly the master at the default location along all the axes).
That one defines the variable font's final glyph set because the deltas are added up to that master font. If non-default masters do not have a given glyph, then that master does not participate and no deltas are added for it.

Are you suggesting that we warn if glyphs that are named the same have different unicodes? It seems unlikely to me, and probably a font bug, which a linter could take care of before building.

For sure we don't want to warn if a non-default master do not have glyphs that are present in the default, because that's "by design" to mean sparse masters. Potentially one could warn if a non-default master has more glyphs than the default master, which will be ignoreed and dropped in the final VF.

> a fontmake problem or a ufo2ft problem or a fontTools.varLib.merge problem

kinda both a fontmake.instantiator problem (which creates static instance UFOs) and a varLib problem (which creates VFs)

**Comment by simoncozens** [comment]
When I reported this it was a theoretical possibility for which I created some test fonts. But I've since found it occurring in the wild (https://github.com/TypeNetwork/Amstelvar/issues/184), which demonstrated a real bug - because a misnamed glyph meant that a glyph was not present in some masters, Amstelvar was getting "sparse master" glyphs instead of the designed versions. And nobody knew because there was no warning. Maybe a linter should take care of it, but *is* there a linter?

So I think there should probably at least be a warning for this situation.

**Comment by anthrotype** [comment]
so the problem is how do we distinguish an unintentionally mis-named glyph in a non-default master from one that is missing from an intentionally sparse master?

**Comment by anthrotype** [comment]
ok, we could take the the glyph set of the default master as the ideal superset of all the non-default masters glyph sets; if any non default master contain more glyphs than the default master, we warn. If it contain less, we don't warn and we assume that's intentionally left sparse. WDYT?

**Comment by simoncozens** [comment]
Exactly. Any glyphs in a non-default that aren't in the default are probably bugs.

**Comment by simoncozens** [comment]
I think what I'll do is write a UFO/DS linting library, which fontmake can call.

**Comment by simoncozens** [comment]
Or maybe it's better as a fontbakery thing. I don't know. I'll write the code anyway, then we can work out where to put it.

**Comment by simoncozens** [comment]
https://github.com/googlefonts/fontbakery/pull/3168
